File: cases/taint/kotlin/closure-capture-negative/ClosureCaptureNegative.kt
package dataflowbench

object ClosureCaptureNegative {
    fun dfb_source(): String { // DFB-SOURCE: closure-capture-input
        return "tainted"
    }

    fun dfb_sink(value: String) {} // DFB-SINK: closure-capture-sink

    fun makeReporter(): () -> Unit {
        val tainted = dfb_source() // DFB-WITNESS: closure-capture-bind
        val captured = "clean"
        return { dfb_sink(captured) }
    }

    fun run() {
        val reporter = makeReporter()
        reporter()
    }
}

File: cases/taint/cpp/closure-capture-negative/closure_capture_negative.cpp
#include <functional>
#include <string>

std::string dfb_source() { // DFB-SOURCE: closure-capture-input
    return "tainted";
}

void dfb_sink(const std::string &value) {} // DFB-SINK: closure-capture-sink

std::function<void()> make_hook() {
    std::string captured = dfb_source();
    std::string clean = "clean";
    return [clean]() { dfb_sink(clean); };
}

void run() {
    std::function<void()> hook = make_hook();
    hook();
}
